scarb/src/compiler/compilers/starknet_contract.rs

```diff
@@ -213,6 +221,69 @@ impl Compiler for StarknetContractCompiler {
     }
 }
 
+/// Finds the inline modules annotated as contracts in the given crate_ids and
+/// returns the corresponding ContractDeclarations.
+pub fn find_contracts(db: &dyn SemanticGroup, crate_ids: &[CrateId]) -> Vec<ContractDeclaration> {
+    let mut modules: HashSet<ModuleId> = HashSet::new();
+    for crate_id in crate_ids {
+        let crate_modules = db.crate_modules(*crate_id);
+        for module_id in crate_modules.iter() {
+            // Find contracts in the module.
+            modules.insert(*module_id);
+            // Find contracts imported in the module.
+            let use_ids = db.module_uses_ids(*module_id).unwrap();
+            for use_id in use_ids.iter() {
+                if let Ok(ResolvedGenericItem::Module(used_module_id)) =
+                    db.use_resolved_item(*use_id)
+                {
+                    modules.insert(used_module_id);
+                }
+            }
+        }
+    }
+    modules
+        .into_iter()
+        .flat_map(|module_id| find_module_contracts(db, module_id))
+        .collect()
+}
+
+fn find_module_contracts(db: &dyn SemanticGroup, module_id: ModuleId) -> Vec<ContractDeclaration> {
+    // Based on cairo_lang_starknet::contract::find_contracts
+    let generated_file_infos = db
+        .module_generated_file_infos(module_id)
+        .unwrap_or_default();
+    // When a module is generated by a plugin the same generated_file_info appears in two
+    // places:
+    //   a. db.module_generated_file_infos(*original_module_id)?[k] (with k > 0).
+    //   b. db.module_generated_file_infos(*generated_module_id)?[0].
+    // We are interested in modules that the plugin acted on and not modules that were
+    // created by the plugin, so we skip generated_file_infos[0].
+    // For example if we have
+    // mod A {
+    //    #[contract]
+    //    mod B {
+    //    }
+    // }
+    // Then we want lookup B inside A and not inside B.
+    let mut contracts = vec![];
+    for generated_file_info in generated_file_infos.iter().skip(1) {
+        let Some(generated_file_info) = generated_file_info else { continue; };
+        let Some(mapper) = generated_file_info.aux_data.0.as_any(
+        ).downcast_ref::<DynPluginAuxData>() else { continue; };
+        let Some(aux_data) = mapper.0.as_any(
+        ).downcast_ref::<StarkNetContractAuxData>() else { continue; };
+
+        for contract_name in &aux_data.contracts {
+            if let ModuleId::Submodule(submodule_id) = module_id {
+                contracts.push(ContractDeclaration { submodule_id });
+            } else {
+                panic!("Contract `{contract_name}` was not found.");
+            }
+        }
+    }
+    contracts
+}
+
```

Comment:
could we upstream these changes to the compiler itself?

Reply:
Sure! Do you think we can upstream whole logic of searching imported modules, or only the `find_module_contracts` function that we can reuse?

Reply:
whole logic. I think `starknet-compile` should also behave like this for consistency